### PR TITLE
Prepare checklist for system development

### PR DESCRIPTION
**Pull request by JohnShiuMK**
Hi @ttimbers,

As discussed last Friday, we want to narrow our focus to a concise set of checklist items for system development. Here is the checklist in HTML format: [Checklist for System Development](https://github.com/UBC-MDS/test-creation/blob/prep-checklist-for-sys-dev/checklist/checklist_sys.html).

We have prepared the following items:
 - 2.1 Test Data Fetching and File Reading
 - 3.1 Validate Data Shape and Values
 - 3.2 Check for Duplicate Records in Data
 - 4.1 Verify Data Split Proportion
 - 5.1 Test Model Output Shape
 - 6.1 Verify Evaluation Metrics Implementation
 - 6.2 Evaluate Model’s Performance Against Thresholds
 - 8.1 Validate Outliers Detection and Handling

We believe these items are manageable in terms of granularity for system development and evaluation. **Do you think they are granular enough, or do they appear too broad?**

Additionally, **we would appreciate any feedback on our HTML visualization.**

Specifically for the last test case, "8.1 Validate Outliers Detection and Handling", **do you know of any test case examples or projects that include test functions for handling outliers in input data?**

Thank you for your assistance.

### CODE REVIEW

**Comment by jinyz8888** [comment]
Hello @JohnShiuMK I've reviewed the checklist at this page and compared it with our current version. Your version has 4 items overlapped with the existing checklist items, please check https://github.com/UBC-MDS/test-creation/tree/main/checklist/checklist.csv. 

I introduced to your four new items in detail at noon on May 21. please give them a chance to be discussed. It is okay if they are not suitable for our system. For the future, I recommend we discuss our findings and align our checklists before finalizing any new entries to ensure consistency and completeness!

**Comment by JohnShiuMK** [comment]
> Hello @JohnShiuMK I've reviewed the checklist at this page and compared it with our current version. Your version has 4 items overlapped with the existing checklist items, please check https://github.com/UBC-MDS/test-creation/tree/main/checklist/checklist.csv.
> 
> I introduced to your four new items in detail at noon on May 21. please give them a chance to be discussed. It is acceptable if finally we find that they are not suitable for our system. For the future, I recommend we discuss our findings and align our checklists before finalizing any new entries to ensure consistency and completeness!

Sorry @jinyz8888, maybe I have misunderstood your "Yes, I’m okay with it". I thought you're also okay with what I mentioned in the first paragraph: "I have added some items related to the Machine learning pipeline testing to the list." 
Screenshot 2024-05-22 at 11 36 13 PM

I admit that I should have introduced you the items. I will make myself clear and make sure the alignment is there next time. Thanks for bringing it up and sorry again for the mis-communication.

File: checklist/checklist_sys.html [hunk 1]
```diff
@@ -0,0 +1,155 @@
+<h1 id="checklist-for-tests-in-machine-learning-projects">Checklist for
+Tests in Machine Learning Projects</h1>
+<p><strong>Description</strong>: This is a comprehensive checklist for
+evaluating the data and ML pipeline based on identified testing
+strategies from experts in the field.</p>
+<h2 id="general">1 General</h2>
```
Comment (ttimbers) [inline checklist/checklist_sys.html:6]: Why does the General header have a "1"? It is not a checklist item. This is a bit confusing to me.

File: checklist/checklist_sys.csv/tests.csv [hunk 2]
```diff
@@ -0,0 +1,9 @@
+ID,Topic,Title,Requirement,Explanation,References
+2.1,Data Presence,Test Data Fetching and File Reading,"Verify that the data fetching API or data file reading functionality works correctly. Ensure that proper error handling is in place for scenarios such as missing files, incorrect file formats, and network errors.","Ensure that the code responsible for fetching or reading data can handle errors. This means if the file is missing, the format is wrong, or there's a network issue, the system should not crash but should provide a clear error message indicating the problem.",(general knowledge)
```
Comment (ttimbers) [inline checklist/checklist_sys.csv/tests.csv:2]: Under the topic of "data presence" (or saving and loading data) - At the beginning of a data analysis pipeline I would expect two possible functions: 

1. A data retrieval function (e.g., downloads data files from a url, reads data from a database, requests data from an API, screen scrapes data from a website).

2. A saving/writing the retrieved data to disc function, which takes the data retrieved into memory and writes it to disc.

These might be general knowledge, but these were also written in the Microsoft guide, and so I think that should be listed as a reference. *Note - not all projects may have these functions. For example some projects data might have been collected manually.*

I think for all test suites, for any function, I would expect the following three buckets of test cases:
- Simple expected use cases
- Edge cases (unexpected, or rare use cases)
- Errors

File: checklist/checklist_sys.csv/tests.csv [hunk 3]
```diff
@@ -0,0 +1,9 @@
+ID,Topic,Title,Requirement,Explanation,References
+2.1,Data Presence,Test Data Fetching and File Reading,"Verify that the data fetching API or data file reading functionality works correctly. Ensure that proper error handling is in place for scenarios such as missing files, incorrect file formats, and network errors.","Ensure that the code responsible for fetching or reading data can handle errors. This means if the file is missing, the format is wrong, or there's a network issue, the system should not crash but should provide a clear error message indicating the problem.",(general knowledge)
+3.1,Data Quality,Validate Data Shape and Values,"Check that the data has the expected shape and that all values meet domain-specific constraints, such as non-negative distances.","Check that the data being used has the correct structure (like having the right number of columns) and that the values within the data make sense (e.g., distances should not be negative). This ensures that the data is valid and reliable for model training.","alexander2024Evaluating, ISO/IEC5259"
```
Comment (ttimbers) [inline checklist/checklist_sys.csv/tests.csv:3]: This does seem important, but I wonder does this kind of test go in the `tests` folder, or is this part of the EDA script in a data analysis? This seems a different class of tests compared to the tests above which are testing the code for retrieving and saving data. This makes me think about the test pyramid in Openja et al... This is not a negative comment, just a question to drive future discussion.

File: checklist/checklist_sys.csv/tests.csv [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+3.1,Data Quality,Validate Data Shape and Values,"Check that the data has the expected shape and that all values meet domain-specific constraints, such as non-negative distances.","Check that the data being used has the correct structure (like having the right number of columns) and that the values within the data make sense (e.g., distances should not be negative). This ensures that the data is valid and reliable for model training.","alexander2024Evaluating, ISO/IEC5259"
+3.2,Data Quality,Check for Duplicate Records in Data,Check for duplicate records in the dataset and ensure that there are none.,"Ensure that the dataset does not contain duplicate entries, as these can skew the results and reduce the model's performance. The test should identify any repeated records so they can be removed or investigated.",ISO/IEC5259
```
Comment (ttimbers) [inline checklist/checklist_sys.csv/tests.csv:4]: Same comment here as for test checklist item 3.1.

File: checklist/checklist_sys.csv/tests.csv [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+3.2,Data Quality,Check for Duplicate Records in Data,Check for duplicate records in the dataset and ensure that there are none.,"Ensure that the dataset does not contain duplicate entries, as these can skew the results and reduce the model's performance. The test should identify any repeated records so they can be removed or investigated.",ISO/IEC5259
+4.1,Data Ingestion,Verify Data Split Proportion,Check that the data is split into training and testing sets in the expected proportion.,"Confirm that the data is divided correctly into training and testing sets according to the intended ratio. This is crucial for ensuring that the model is trained and evaluated properly, with representative samples in each set.","openja2023studying, DBLP:conf/recsys/Kula15, singh2020mmf"
```
Comment (ttimbers) [inline checklist/checklist_sys.csv/tests.csv:5]: So here, in a data splitting script, I think I would expect at least three functions:

1. A function to read in the raw data that exists on disc.
2. A function to split the data into at least a training and test set (maybe a training, validation and test set). This function might also stratify across important variables (most often the target, but there can be others) as well as group data for appropriate splitting of non-independent observations (e.g., in the case of time series or spatial data).
3. A function that writes the raw trained data and raw test data to disc.

Oftentimes such a script at this point in the ML pipeline would also create a pre-processor (scikit-learn) pipeline object from the training data, which would also be saved to disc. It is useful to do this early on, so that one can pre-process data to be used for EDA, as well as having it available to read into later training and test/prediction scripts.

File: checklist/checklist_sys.csv/tests.csv [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+4.1,Data Ingestion,Verify Data Split Proportion,Check that the data is split into training and testing sets in the expected proportion.,"Confirm that the data is divided correctly into training and testing sets according to the intended ratio. This is crucial for ensuring that the model is trained and evaluated properly, with representative samples in each set.","openja2023studying, DBLP:conf/recsys/Kula15, singh2020mmf"
+5.1,Model Fitting,Test Model Output Shape,Validate that the model's output has the expected shape.,"Ensure that the output from the model has the correct dimensions and structure. For example, in a classification task, if the model should output probabilities for each class, the test should verify that the output is an array with the correct dimensions. Ensuring the correct output shape helps prevent runtime errors and ensures consistency in how data is handled downstream.","openja2023studying, DBLP:conf/recsys/Kula15, singh2020mmf"
```
Comment (ttimbers) [inline checklist/checklist_sys.csv/tests.csv:6]: So I imagine a test like this would map to a function in a the training/fitting script which took in at a minimum, the pre-proccessor pipeline object, and raw training data. In some cases it might also take in some parameters for the models and fitting procedures (e.g., cross-validation)? It would return the fitted model (scikit-learn) pipeline object?

**Review by ttimbers — COMMENTED** [review]
Here is some feedback on the first 5 checklist items. We can maybe use this as some starting points for discussion in our meeting this week?

File: checklist/checklist_sys.csv/tests.csv (hunk 3 above)
Comment (tonyshumlh) [reply checklist/checklist_sys.csv/tests.csv:3]: Hi Tiffany. 
I think EDA and data wrangling are to explore and transform the data, 
while the test case here is to validate the raw/processed data after a bunch of operations, to ensure the data is as expected for the next stage (e.g. model training)

File: checklist/checklist_sys.csv/tests.csv (hunk 5 above)
Comment (tonyshumlh) [reply checklist/checklist_sys.csv/tests.csv:5]: As the functions for building the ML model, I would expect the 3 function you mentioned
As a function for unit case, I can imagine a test function that read in train & test data (e.g. from `data/processed/`) and verify if the size ratio is as expected (e.g. `train_data.shape[0]/test_data.shape[0] == 0.8/0.2`)

File: checklist/checklist_sys.csv/tests.csv (hunk 2 above)
Comment (tonyshumlh) [reply checklist/checklist_sys.csv/tests.csv:2]: > I think for all test suites, for any function, I would expect the following three buckets of test cases:
> Simple expected use cases
> Edge cases (unexpected, or rare use cases)
> Errors
Agree

File: checklist/checklist_sys.csv/tests.csv (hunk 6 above)
Comment (tonyshumlh) [reply checklist/checklist_sys.csv/tests.csv:6]: I think it is a model pre-train test without needing trained parameters.
I can image a unit test for this would take in a model (initialized but not fitted) and a small portion of training data, fit the model and verify the shape of the `prediction`
e.g. if the expected output is the class label, the shape should be `N * 1` where N is the #row in training data

**Comment by tonyshumlh** [comment]
> 6.1 Verify Evaluation Metrics Implementation
> 6.2 Evaluate Model’s Performance Against Thresholds
> 8.1 Validate Outliers Detection and Handling (Optional)
Hi @ttimbers 
Would be great if we can have some comments on these items too. We can then adjust the checklist items based on our discussion. Thanks a lot :)

File: checklist/checklist_sys.csv/tests.csv (hunk 3 above)
Comment (tonyshumlh) [reply checklist/checklist_sys.csv/tests.csv:3]: Hi @ttimbers 
Indeed we also have a question whether this Checklist item shall target raw data or processed data?
To me, targeting raw data implies a validation on the data source, while targeting processed data implies a validation on the pre-processing operation by the user.

**Comment by JohnShiuMK** [comment]
to be developed after 591
